Switch to standard OpenSearch gradle build

peternied opened the pull request:
### Description
By switching to the standard build process we should no longer have to solve build related issues in isolation, we should be able to reuse all the standard scripts and practices by the OpenSearch organization.


* Rewrote build.gradle to follow OpenSearch plugin standards
  * Enabled license headers check on the repository
  * Did not enable several new repository checks
  * Added maven publishing for security
  * Removed excess forced dependency resolutions
  * Rebuilt projects dependencies, release plugin binary has the exact same dependencies and versions
  * Converted dependencies into runtime dependencies to avoid use during coding
  * Converted dependencies from project wide to test only dependencies
  * jackson-databind version comes from OpenSearch
  * Replaced handmade build manifest with git properties to automated version used by OpenSearch
* Added license headers to files that were missing them
* Checkstyle improvements
  * Moved checkstyle file into directory off of the project root
* Moved configuration files into config directory off of project root
* Using default test runner from OpenSearch
  * Switched to RandomizedTest as the base test class
  * Parameterized test runner cannot be used with RandomizedTest converted existing tests
  * Fixed tests resource issues, added new fields to make this consistent in the codebase
  * Fixed issue with leaky environment variable for security root directory, refactored usage
  * Removed unneeded setDefaultUncaughtExceptionHandler in tests
  * Fixed issues with deprecated internal reflection from Mockito
  * Disabled ThreadLeak detection as it is catching real issues that are mitigated by retries
* Initial pass on non-inclusive terminology, commented out exclusions
* Removed test dependency on scala
* Removed test dependency on legacy xmlsecurity library

### Issues Resolved
* Resolves #1816
* Resolves #1839
* Resolves #1801

### Testing

<details><summary>Security plugin - comparison</summary>
<p>

```
Folder Compare
Produced: 6/16/2022 3:28:53 PM

Mode:  All
Left base folder: C:\Users\petern\distros\opensearch-security-2.1.0.0-SNAPSHOT.zip
Right base folder: C:\Users\petern\distros\opensearch-security-2.1.0.0-SNAPSHOT-newest.zip
Name                                                           Size      Modified                 Name                                                           Size      Modified
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
config                                                            39,560 6/16/2022 5:59:40 PM     config                                                            39,560 6/16/2022 6:08:20 PM
tools                                                             26,656 6/16/2022 5:59:40 PM     tools                                                             26,656 6/16/2022 6:08:20 PM
accessors-smart-2.4.7.jar                                         29,489 5/11/2022 4:43:28 PM  =  accessors-smart-2.4.7.jar                                         29,489 5/11/2022 4:43:28 PM
aggs-matrix-stats-client-2.1.0-SNAPSHOT.jar                       57,358 6/15/2022 10:52:56 PM =  aggs-matrix-stats-client-2.1.0-SNAPSHOT.jar                       57,358 6/15/2022 10:52:56 PM
asm-9.1.jar                                                      121,790 5/11/2022 4:43:28 PM  =  asm-9.1.jar                                                      121,790 5/11/2022 4:43:28 PM
bcprov-jdk15on-1.67.jar                                        5,961,136 5/11/2022 4:43:28 PM  =  bcprov-jdk15on-1.67.jar                                        5,961,136 5/11/2022 4:43:28 PM
checker-qual-3.5.0.jar                                           214,381 5/11/2022 4:43:28 PM  =  checker-qual-3.5.0.jar                                           214,381 5/11/2022 4:43:28 PM
commons-cli-1.3.1.jar                                             52,988 5/11/2022 4:43:28 PM  =  commons-cli-1.3.1.jar                                             52,988 5/11/2022 4:43:28 PM
commons-codec-1.14.jar                                           347,669 5/11/2022 4:43:28 PM  =  commons-codec-1.14.jar                                           347,669 5/11/2022 4:43:28 PM
commons-collections-3.2.2.jar                                    588,337 5/11/2022 4:43:28 PM  =  commons-collections-3.2.2.jar                                    588,337 5/11/2022 4:43:28 PM
commons-lang-2.4.jar                                             261,809 5/11/2022 4:43:28 PM  =  commons-lang-2.4.jar                                             261,809 5/11/2022 4:43:28 PM
commons-lang3-3.4.jar                                            434,678 5/11/2022 4:43:28 PM  =  commons-lang3-3.4.jar                                            434,678 5/11/2022 4:43:28 PM
commons-logging-1.2.jar                                           61,829 5/11/2022 4:43:28 PM  =  commons-logging-1.2.jar                                           61,829 5/11/2022 4:43:28 PM
commons-text-1.2.jar                                             136,544 5/11/2022 4:43:28 PM  =  commons-text-1.2.jar                                             136,544 5/11/2022 4:43:28 PM
compiler-0.9.10.jar                                              111,518 5/11/2022 4:43:28 PM  =  compiler-0.9.10.jar                                              111,518 5/11/2022 4:43:28 PM
cryptacular-1.2.4.jar                                            162,471 5/11/2022 4:43:28 PM  =  cryptacular-1.2.4.jar                                            162,471 5/11/2022 4:43:28 PM
cxf-core-3.4.5.jar                                             1,434,086 5/11/2022 4:43:28 PM  =  cxf-core-3.4.5.jar                                             1,434,086 5/11/2022 4:43:28 PM
cxf-rt-rs-json-basic-3.4.5.jar                                    22,028 5/11/2022 4:43:28 PM  =  cxf-rt-rs-json-basic-3.4.5.jar                                    22,028 5/11/2022 4:43:28 PM
cxf-rt-rs-security-jose-3.4.5.jar                                234,129 5/11/2022 4:43:28 PM  =  cxf-rt-rs-security-jose-3.4.5.jar                                234,129 5/11/2022 4:43:28 PM
cxf-rt-security-3.4.5.jar                                         41,621 5/11/2022 4:43:28 PM  =  cxf-rt-security-3.4.5.jar                                         41,621 5/11/2022 4:43:28 PM
error_prone_annotations-2.3.4.jar                                 13,879 5/11/2022 4:43:28 PM  =  error_prone_annotations-2.3.4.jar                                 13,879 5/11/2022 4:43:28 PM
eventbus-3.2.0.jar                                                60,100 5/11/2022 4:43:28 PM  =  eventbus-3.2.0.jar                                                60,100 5/11/2022 4:43:28 PM
failureaccess-1.0.1.jar                                            4,617 5/11/2022 4:43:28 PM  =  failureaccess-1.0.1.jar                                            4,617 5/11/2022 4:43:28 PM
guava-30.0-jre.jar                                             2,858,426 5/11/2022 4:43:28 PM  =  guava-30.0-jre.jar                                             2,858,426 5/11/2022 4:43:28 PM
httpasyncclient-4.1.4.jar                                        179,964 5/11/2022 4:43:28 PM  =  httpasyncclient-4.1.4.jar                                        179,964 5/11/2022 4:43:28 PM
httpclient-4.5.13.jar                                            780,321 5/11/2022 4:43:28 PM  =  httpclient-4.5.13.jar                                            780,321 5/11/2022 4:43:28 PM
httpclient-cache-4.5.13.jar                                      164,090 5/11/2022 4:43:28 PM  =  httpclient-cache-4.5.13.jar                                      164,090 5/11/2022 4:43:28 PM
httpcore-4.4.12.jar                                              328,347 5/11/2022 4:43:28 PM  =  httpcore-4.4.12.jar                                              328,347 5/11/2022 4:43:28 PM
httpcore-nio-4.4.12.jar                                          367,672 5/11/2022 4:43:28 PM  =  httpcore-nio-4.4.12.jar                                          367,672 5/11/2022 4:43:28 PM
istack-commons-runtime-3.0.12.jar                                 29,807 5/11/2022 4:43:28 PM  =  istack-commons-runtime-3.0.12.jar                                 29,807 5/11/2022 4:43:28 PM
j2objc-annotations-1.3.jar                                         8,781 5/11/2022 4:43:28 PM  =  j2objc-annotations-1.3.jar                                         8,781 5/11/2022 4:43:28 PM
jackson-annotations-2.13.2.jar                                    75,717 5/11/2022 4:43:28 PM  =  jackson-annotations-2.13.2.jar                                    75,717 5/11/2022 4:43:28 PM
                                                                                               << jackson-databind-2.13.2.2.jar                                  1,535,087 5/13/2022 9:39:48 PM
jackson-databind-2.13.2.jar                                    1,534,715 5/11/2022 4:43:28 PM  >>
jakarta.activation-1.2.2.jar                                      68,453 5/11/2022 4:43:34 PM  =  jakarta.activation-1.2.2.jar                                      68,453 5/11/2022 4:43:34 PM
jakarta.annotation-api-1.3.5.jar                                  25,058 5/11/2022 4:43:28 PM  =  jakarta.annotation-api-1.3.5.jar                                  25,058 5/11/2022 4:43:28 PM
jakarta.xml.bind-api-2.3.3.jar                                   115,638 5/11/2022 4:43:28 PM  =  jakarta.xml.bind-api-2.3.3.jar                                   115,638 5/11/2022 4:43:28 PM
java-saml-2.5.0.jar                                               12,285 5/11/2022 4:43:28 PM  =  java-saml-2.5.0.jar                                               12,285 5/11/2022 4:43:28 PM
java-saml-core-2.5.0.jar                                         113,170 5/11/2022 4:43:28 PM  =  java-saml-core-2.5.0.jar                                         113,170 5/11/2022 4:43:28 PM
java-support-7.5.1.jar                                           272,381 5/11/2022 4:43:28 PM  =  java-support-7.5.1.jar                                           272,381 5/11/2022 4:43:28 PM
jaxb-runtime-2.3.4.jar                                         1,037,093 5/11/2022 4:43:28 PM  =  jaxb-runtime-2.3.4.jar                                         1,037,093 5/11/2022 4:43:28 PM
jjwt-api-0.10.8.jar                                               74,746 5/11/2022 4:43:28 PM  =  jjwt-api-0.10.8.jar                                               74,746 5/11/2022 4:43:28 PM
jjwt-impl-0.10.8.jar                                              70,159 5/11/2022 4:43:34 PM  =  jjwt-impl-0.10.8.jar                                              70,159 5/11/2022 4:43:34 PM
jjwt-jackson-0.10.8.jar                                            4,714 5/11/2022 4:43:34 PM  =  jjwt-jackson-0.10.8.jar                                            4,714 5/11/2022 4:43:34 PM
json-flattener-0.5.0.jar                                          25,211 5/11/2022 4:43:28 PM  =  json-flattener-0.5.0.jar                                          25,211 5/11/2022 4:43:28 PM
json-path-2.4.0.jar                                              223,186 5/11/2022 4:43:28 PM  =  json-path-2.4.0.jar                                              223,186 5/11/2022 4:43:28 PM
json-smart-2.4.7.jar                                             119,227 5/11/2022 4:43:28 PM  =  json-smart-2.4.7.jar                                             119,227 5/11/2022 4:43:28 PM
jsr305-3.0.2.jar                                                  19,936 5/11/2022 4:43:24 PM  =  jsr305-3.0.2.jar                                                  19,936 5/11/2022 4:43:24 PM
kafka-clients-3.0.0.jar                                        4,855,638 5/11/2022 4:43:28 PM  =  kafka-clients-3.0.0.jar                                        4,855,638 5/11/2022 4:43:28 PM
lang-mustache-client-2.1.0-SNAPSHOT.jar                           62,947 6/15/2022 10:52:56 PM =  lang-mustache-client-2.1.0-SNAPSHOT.jar                           62,947 6/15/2022 10:52:56 PM
ldaptive-1.2.3.jar                                               802,456 5/11/2022 4:43:28 PM  =  ldaptive-1.2.3.jar                                               802,456 5/11/2022 4:43:28 PM
listenablefuture-9999.0-empty-to-avoid-conflict-with-guava.jar     2,199 5/11/2022 4:43:28 PM  =  listenablefuture-9999.0-empty-to-avoid-conflict-with-guava.jar     2,199 5/11/2022 4:43:28 PM
lz4-java-1.7.1.jar                                               649,950 5/11/2022 4:43:34 PM  =  lz4-java-1.7.1.jar                                               649,950 5/11/2022 4:43:34 PM
mapper-extras-client-2.1.0-SNAPSHOT.jar                           83,904 6/15/2022 10:52:32 PM =  mapper-extras-client-2.1.0-SNAPSHOT.jar                           83,904 6/15/2022 10:52:32 PM
metrics-core-3.1.2.jar                                           112,558 5/11/2022 4:43:28 PM  =  metrics-core-3.1.2.jar                                           112,558 5/11/2022 4:43:28 PM
minimal-json-0.9.5.jar                                            34,221 5/11/2022 4:43:28 PM  =  minimal-json-0.9.5.jar                                            34,221 5/11/2022 4:43:28 PM
netty-buffer-4.1.73.Final.jar                                    302,927 5/11/2022 4:43:28 PM  =  netty-buffer-4.1.73.Final.jar                                    302,927 5/11/2022 4:43:28 PM
netty-codec-4.1.73.Final.jar                                     337,170 5/11/2022 4:43:28 PM  =  netty-codec-4.1.73.Final.jar                                     337,170 5/11/2022 4:43:28 PM
netty-codec-http-4.1.73.Final.jar                                634,551 5/11/2022 4:43:28 PM  =  netty-codec-http-4.1.73.Final.jar                                634,551 5/11/2022 4:43:28 PM
netty-common-4.1.73.Final.jar                                    649,516 5/11/2022 4:43:28 PM  =  netty-common-4.1.73.Final.jar                                    649,516 5/11/2022 4:43:28 PM
netty-handler-4.1.73.Final.jar                                   522,638 5/11/2022 4:43:28 PM  =  netty-handler-4.1.73.Final.jar                                   522,638 5/11/2022 4:43:28 PM
netty-resolver-4.1.73.Final.jar                                   37,776 5/11/2022 4:43:28 PM  =  netty-resolver-4.1.73.Final.jar                                   37,776 5/11/2022 4:43:28 PM
netty-transport-4.1.73.Final.jar                                 478,804 5/11/2022 4:43:28 PM  =  netty-transport-4.1.73.Final.jar                                 478,804 5/11/2022 4:43:28 PM
opensaml-core-3.4.5.jar                                          181,870 5/11/2022 4:43:28 PM  =  opensaml-core-3.4.5.jar                                          181,870 5/11/2022 4:43:28 PM
opensaml-messaging-api-3.4.5.jar                                  56,249 5/11/2022 4:43:28 PM  =  opensaml-messaging-api-3.4.5.jar                                  56,249 5/11/2022 4:43:28 PM
opensaml-profile-api-3.4.5.jar                                    31,118 5/11/2022 4:43:28 PM  =  opensaml-profile-api-3.4.5.jar                                    31,118 5/11/2022 4:43:28 PM
opensaml-saml-api-3.4.5.jar                                      407,731 5/11/2022 4:43:28 PM  =  opensaml-saml-api-3.4.5.jar                                      407,731 5/11/2022 4:43:28 PM
opensaml-saml-impl-3.4.5.jar                                   1,169,221 5/11/2022 4:43:28 PM  =  opensaml-saml-impl-3.4.5.jar                                   1,169,221 5/11/2022 4:43:28 PM
opensaml-security-api-3.4.5.jar                                   66,957 5/11/2022 4:43:28 PM  =  opensaml-security-api-3.4.5.jar                                   66,957 5/11/2022 4:43:28 PM
opensaml-security-impl-3.4.5.jar                                 112,079 5/11/2022 4:43:28 PM  =  opensaml-security-impl-3.4.5.jar                                 112,079 5/11/2022 4:43:28 PM
opensaml-soap-api-3.4.5.jar                                      169,987 5/11/2022 4:43:28 PM  =  opensaml-soap-api-3.4.5.jar                                      169,987 5/11/2022 4:43:28 PM
opensaml-soap-impl-3.4.5.jar                                     403,255 5/11/2022 4:43:28 PM  =  opensaml-soap-impl-3.4.5.jar                                     403,255 5/11/2022 4:43:28 PM
opensaml-storage-api-3.4.5.jar                                    23,813 5/11/2022 4:43:28 PM  =  opensaml-storage-api-3.4.5.jar                                    23,813 5/11/2022 4:43:28 PM
opensaml-xmlsec-api-3.4.5.jar                                    209,519 5/11/2022 4:43:28 PM  =  opensaml-xmlsec-api-3.4.5.jar                                    209,519 5/11/2022 4:43:28 PM
opensaml-xmlsec-impl-3.4.5.jar                                   297,343 5/11/2022 4:43:28 PM  =  opensaml-xmlsec-impl-3.4.5.jar                                   297,343 5/11/2022 4:43:28 PM
opensearch-rest-client-2.1.0-SNAPSHOT.jar                         65,600 6/15/2022 10:52:32 PM =  opensearch-rest-client-2.1.0-SNAPSHOT.jar                         65,600 6/15/2022 10:52:32 PM
opensearch-rest-high-level-client-2.1.0-SNAPSHOT.jar             307,667 6/15/2022 10:52:32 PM =  opensearch-rest-high-level-client-2.1.0-SNAPSHOT.jar             307,667 6/15/2022 10:52:32 PM
opensearch-security-2.1.0.0-SNAPSHOT.jar                       1,170,360 6/16/2022 5:59:38 PM  << opensearch-security-2.1.0.0-SNAPSHOT.jar                       1,174,435 6/16/2022 6:08:18 PM
parent-join-client-2.1.0-SNAPSHOT.jar                             87,032 6/15/2022 10:52:56 PM =  parent-join-client-2.1.0-SNAPSHOT.jar                             87,032 6/15/2022 10:52:56 PM
plugin-descriptor.properties                                         216 6/16/2022 5:59:36 PM  << plugin-descriptor.properties                                       1,951 6/16/2022 6:08:18 PM
plugin-security.policy                                             3,280 6/15/2022 4:15:02 PM  =  plugin-security.policy                                             3,280 6/15/2022 4:15:02 PM
rank-eval-client-2.1.0-SNAPSHOT.jar                               82,422 6/15/2022 10:52:56 PM =  rank-eval-client-2.1.0-SNAPSHOT.jar                               82,422 6/15/2022 10:52:56 PM
slf4j-api-1.7.30.jar                                              41,472 5/11/2022 4:43:24 PM  =  slf4j-api-1.7.30.jar                                              41,472 5/11/2022 4:43:24 PM
snappy-java-1.1.8.1.jar                                        1,952,778 5/11/2022 4:43:34 PM  =  snappy-java-1.1.8.1.jar                                        1,952,778 5/11/2022 4:43:34 PM
stax2-api-4.2.1.jar                                              195,909 5/11/2022 4:43:28 PM  =  stax2-api-4.2.1.jar                                              195,909 5/11/2022 4:43:28 PM
transport-netty4-client-2.1.0-SNAPSHOT.jar                        73,277 6/15/2022 10:52:56 PM =  transport-netty4-client-2.1.0-SNAPSHOT.jar                        73,277 6/15/2022 10:52:56 PM
txw2-2.3.4.jar                                                    71,975 5/11/2022 4:43:28 PM  =  txw2-2.3.4.jar                                                    71,975 5/11/2022 4:43:28 PM
woodstox-core-6.2.6.jar                                        1,584,445 5/11/2022 4:43:28 PM  =  woodstox-core-6.2.6.jar                                        1,584,445 5/11/2022 4:43:28 PM
xmlschema-core-2.2.5.jar                                         173,293 5/11/2022 4:43:28 PM  =  xmlschema-core-2.2.5.jar                                         173,293 5/11/2022 4:43:28 PM
xmlsec-2.2.3.jar                                               1,123,199 5/11/2022 4:43:28 PM  =  xmlsec-2.2.3.jar                                               1,123,199 5/11/2022 4:43:28 PM
zjsonpatch-0.4.4.jar                                              30,476 5/11/2022 4:43:28 PM  =  zjsonpatch-0.4.4.jar                                              30,476 5/11/2022 4:43:28 PM
zstd-jni-1.5.0-2.jar                                           6,762,522 5/11/2022 4:43:34 PM  =  zstd-jni-1.5.0-2.jar                                           6,762,522 5/11/2022 4:43:34 PM
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
```
</p>
</details>

The only meaningful difference seen here is the jackson-databind version, which is in alignment with the OpenSearch repository.

### Check List
- [ ] ~New functionality includes testing~
- [ ] ~New functionality has been documented~
- [x] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

cliu123 commented on build.gradle:
```diff
@@ -9,79 +9,200 @@
  * GitHub history for details.
```
Is this needed?

peternied replied:
This is the standard header that @DarshitChanpura applied in the codebase in https://github.com/opensearch-project/security/pull/1829.  If you'd like to change the standard license, could we have that discussion in a sperate issue?

cliu123 reviewed (dismissed):
These are very sensitive changes. We need to make sure not to change any dependencies, artifacts, tests etc. Any miss could introduce build, test and functionality regressions. A couple of initial comments: 
1. There are only 245 tests running:
`245 tests completed, 1 failed, 8 skipped`
Before the change, 1111 tests running:
`1111 tests completed, 1 failed, 46 skipped`
2. Please verify all the dependencies and artifacts being generated are exactly same.

peternied commented:
> Please verify all the dependencies and artifacts being generated are exactly same.

See in the description `Security plugin - comparison` (it expands with the file diff) the produced artifacts is the same.

> There are only 245 tests running:
245 tests completed, 1 failed, 8 skipped
Before the change, 1111 tests running:
1111 tests completed, 1 failed, 46 skipped

Great idea to check this... I'll see about getting before/after test lists.

cliu123 commented:
> See in the description `Security plugin - comparison` (it expands with the file diff) the produced artifacts is the same.

I see some differences here:
```
                                                                                               << jackson-databind-2.13.2.2.jar                                  1,535,087 5/13/2022 9:39:48 PM
jackson-databind-2.13.2.jar                                    1,534,715 5/11/2022 4:43:28 PM  >>
```
```
opensearch-security-2.1.0.0-SNAPSHOT.jar                       1,170,360 6/16/2022 5:59:38 PM  << opensearch-security-2.1.0.0-SNAPSHOT.jar                       1,174,435 6/16/2022 6:08:18 PM
```
```
plugin-descriptor.properties                                         216 6/16/2022 5:59:36 PM  << plugin-descriptor.properties                                       1,951 6/16/2022 6:08:18 PM

```

peternied commented on build.gradle:
```diff
@@ -9,79 +9,200 @@
- * KIND, either express or implied.  See the License for the
- * specific language governing permissions and limitations
- * under the License.
- */
```
I rewrote this file - this license no longer applies

peternied commented:
@cliu123 Thanks for checking in on those differences, they are all expected, here is a quick rundown with explainations for each.

The jackson-databind version is in now in alignment with the OpenSearch repository, this difference is expected

<details><summary>The **plugin descriptor** follows the template pattern, see expanded diff</summary>
<p>

```
DlsFlsFilterLeafReader.java#L652-L665
Produced: 6/17/2022 1:57:32 PM

Mode:  All, Ignoring Unimportant
Left file: C:\Users\petern\distros\opensearch-security-2.1.0.0-SNAPSHOT.zip\plugin-descriptor.properties
Right file: C:\Users\petern\distros\opensearch-security-2.1.0.0-SNAPSHOT-newest.zip\plugin-descriptor.properties
                                                                   -+ #
                                                                      # SPDX-License-Identifier: Apache-2.0
                                                                      #
                                                                      # The OpenSearch Contributors require contributions made to
                                                                      # this file be licensed under the Apache-2.0 license or a
                                                                      # compatible open source license.
                                                                      #
                                                                      # Modifications Copyright OpenSearch Contributors. See
                                                                      # GitHub history for details.
                                                                      #
-----------------------------------------------------------------------------------------------------------------------------------------------------------
                                                                   =
-----------------------------------------------------------------------------------------------------------------------------------------------------------
                                                                   -+ # OpenSearch plugin descriptor file
                                                                      # This file must exist as 'plugin-descriptor.properties' inside a plugin.
                                                                      #
                                                                      ### example plugin for "foo"
                                                                      #
                                                                      # foo.zip <-- zip file for the plugin, with this structure:
                                                                      # |____   <arbitrary name1>.jar <-- classes, resources, dependencies
                                                                      # |____   <arbitrary nameN>.jar <-- any number of jars
                                                                      # |____   plugin-descriptor.properties <-- example contents below:
                                                                      #
                                                                      # classname=foo.bar.BazPlugin
                                                                      # description=My cool plugin
                                                                      # version=6.0
                                                                      # opensearch.version=6.0
                                                                      # java.version=1.8
                                                                      #
                                                                      ### mandatory elements for all plugins:
                                                                      #
                                                                      # 'description': simple summary of the plugin
-----------------------------------------------------------------------------------------------------------------------------------------------------------
description=Provide access control related features for OpenSearch =  description=Provide access control related features for OpenSearch
-----------------------------------------------------------------------------------------------------------------------------------------------------------
                                                                   -+ #
                                                                      # 'version': plugin's version
-----------------------------------------------------------------------------------------------------------------------------------------------------------
version=2.1.0.0-SNAPSHOT                                           =  version=2.1.0.0-SNAPSHOT
-----------------------------------------------------------------------------------------------------------------------------------------------------------
                                                                   -+ #
                                                                      # 'name': the plugin name
-----------------------------------------------------------------------------------------------------------------------------------------------------------
name=opensearch-security                                           =  name=opensearch-security
-----------------------------------------------------------------------------------------------------------------------------------------------------------
                                                                   -+ #
                                                                      # 'classname': the name of the class to load, fully-qualified
-----------------------------------------------------------------------------------------------------------------------------------------------------------
classname=org.opensearch.security.OpenSearchSecurityPlugin         =  classname=org.opensearch.security.OpenSearchSecurityPlugin
-----------------------------------------------------------------------------------------------------------------------------------------------------------
                                                                   -+ #
                                                                      # 'java.version': version of java the code is built against
                                                                      # use the system property java.specification.version
                                                                      # version string must be a sequence of nonnegative decimal integers
                                                                      # separated by "."'s and may have leading zeros
-----------------------------------------------------------------------------------------------------------------------------------------------------------
java.version=11                                                    =  java.version=11
-----------------------------------------------------------------------------------------------------------------------------------------------------------
                                                                   -+ #
                                                                      # 'opensearch.version': semantic version of opensearch the plugin is compatible with
                                                                      # does not include -SNAPSHOT if compiled against a snapshot build
-----------------------------------------------------------------------------------------------------------------------------------------------------------
opensearch.version=2.1.0                                           =  opensearch.version=2.1.0
-----------------------------------------------------------------------------------------------------------------------------------------------------------
                                                                   -+ #
                                                                      ### optional elements for plugins:
                                                                      #
                                                                      # 'custom.foldername': the custom name of the folder in which the plugin is installed
                                                                      custom.foldername=
                                                                      #
                                                                      # 'extended.plugins': other plugins this plugin extends through SPI
                                                                      extended.plugins=
                                                                      #
                                                                      # 'has.native.controller': whether or not the plugin has a native controller
                                                                      has.native.controller=false
-----------------------------------------------------------------------------------------------------------------------------------------------------------
```
</p></details>


<details><summary>The changes within the **opensearch-security-*.jar** are ignorable, </summary>
<p>

```
Folder Compare
Produced: 6/17/2022 2:02:45 PM

Mode:  All
Ignoring Unimportant
Left base folder: C:\Users\petern\distros\opensearch-security-2.1.0.0-SNAPSHOT\opensearch-security-2.1.0.0-SNAPSHOT.jar
Right base folder: C:\Users\petern\distros\opensearch-security-2.1.0.0-SNAPSHOT-newest\opensearch-security-2.1.0.0-SNAPSHOT.jar
Name                      Size      Modified                Name                      Size      Modified
--------------------------------------------------------------------------------------------------------------------
com                         357,769 6/16/2022 5:59:38 PM    com                         357,769 6/16/2022 6:08:18 PM
\amazon                     357,769 6/16/2022 5:59:38 PM    \amazon                     357,769 6/16/2022 6:08:18 PM
license_headers                 294 6/16/2022 5:59:38 PM    license_headers                 294 6/16/2022 6:08:18 PM
\license_header.txt             294 6/16/2022 5:59:38 PM =  \license_header.txt             294 6/16/2022 6:08:18 PM
META-INF                        389 6/16/2022 5:59:38 PM    META-INF                     12,364 6/16/2022 6:08:18 PM
|                                                        << +LICENSE.txt                 11,357 4/22/2022 5:47:16 PM
\MANIFEST.MF                    389 6/16/2022 5:59:38 PM <> +MANIFEST.MF                    643 6/16/2022 6:08:18 PM
                                                         << \NOTICE.txt                     364 6/7/2022 5:49:38 PM
org                       2,269,860 6/16/2022 5:59:36 PM    org                       2,269,860 6/16/2022 6:08:16 PM
static_config                 8,912 6/16/2022 5:59:38 PM    static_config                 8,912 6/16/2022 6:08:18 PM
+static_action_groups.yml     5,036 6/16/2022 5:59:38 PM =  +static_action_groups.yml     5,036 6/16/2022 6:08:18 PM
+static_audit.yml                18 6/16/2022 5:59:38 PM =  +static_audit.yml                18 6/16/2022 6:08:18 PM
+static_roles.yml             3,714 6/16/2022 5:59:38 PM =  +static_roles.yml             3,714 6/16/2022 6:08:18 PM
\static_tenants.yml             144 6/16/2022 5:59:38 PM =  \static_tenants.yml             144 6/16/2022 6:08:18 PM
git.properties                  572 6/16/2022 5:59:38 PM >>
NOTICE.txt                      364 6/7/2022 5:49:38 PM  =  NOTICE.txt                      364 6/7/2022 5:49:38 PM
THIRD-PARTY.txt               9,864 2/14/2022 7:46:50 PM =  THIRD-PARTY.txt               9,864 2/14/2022 7:46:50 PM
--------------------------------------------------------------------------------------------------------------------

```
</p></details>

<details><summary>Inside **opensearch-security-*.jar**, updated META-INF\MANIFEST.MF file</summary>
<p>

```
Mode:  All, Ignoring Unimportant
Left file: C:\Users\petern\distros\opensearch-security-2.1.0.0-SNAPSHOT\opensearch-security-2.1.0.0-SNAPSHOT.jar\META-INF\MANIFEST.MF
Right file: C:\Users\petern\distros\opensearch-security-2.1.0.0-SNAPSHOT-newest\opensearch-security-2.1.0.0-SNAPSHOT.jar\META-INF\MANIFEST.MF
Manifest-Version: 1.0                                              =  Manifest-Version: 1.0
----------------------------------------------------------------------------------------------------------------------------------------------
Created-By: Gradle 7.4.2                                           <> Module-Origin: git@github.com:opensearch-project/security.git
Build-Jdk: 17.0.1                                                     Change: 3722175d713e0a07b78e6a32cc385d5ed85b44f8
                                                                      X-Compile-OpenSearch-Version: 2.1.0-SNAPSHOT
                                                                      X-Compile-Lucene-Version: 9.2.0-snapshot-ba8c3a8
                                                                      X-Compile-OpenSearch-Snapshot: true
Implementation-Title: OpenSearch Security                             Implementation-Title: org.opensearch#opensearch-security;2.1.0.0-SNAPSHO
                                                                       T
----------------------------------------------------------------------------------------------------------------------------------------------
Implementation-Version: 2.1.0.0-SNAPSHOT                           =  Implementation-Version: 2.1.0.0-SNAPSHOT
----------------------------------------------------------------------------------------------------------------------------------------------
Implementation-Vendor-Id: org.opensearch                           <> Built-Status: integration
Implementation-URL: https://github.com/opensearch-project/security    Built-By: petern
                                                                      Built-OS: Linux
Build-Time: 2022-06-16T17:59:36Z                                      Build-Date: 2022-06-16T18:08:08.170239443Z
Built-By: OpenSearch Security Plugin                                  Gradle-Version: 7.4.2
git-sha1: cd28b115a7ec3b14f465b8f292c98de3709a6322                    Created-By: 17.0.1+12-39 (Oracle Corporation)
                                                                      Build-Java-Version: 17
                                                                      X-Compile-Target-JDK: 11
                                                                      X-Compile-Source-JDK: 11
----------------------------------------------------------------------------------------------------------------------------------------------
                                                                   =
----------------------------------------------------------------------------------------------------------------------------------------------

```
</p></details>

<details><summary>Inside **opensearch-security-*.jar**, deleted git.properties</summary>
<p>

```
git.branch=HEAD
git.build.version=2.1.0.0-SNAPSHOT
git.closest.tag.commit.count=143
git.closest.tag.name=1.1.0.0
git.commit.id=cd28b115a7ec3b14f465b8f292c98de3709a6322
git.commit.id.abbrev=cd28b11
git.commit.id.describe=1.1.0.0-143-gcd28b11
git.commit.message.full=Bump version to 2.1.0.0 (\#1883)\n\nSigned-off-by\: cliu123 <lc12251109@gmail.com>
git.commit.message.short=Bump version to 2.1.0.0 (\#1883)
git.commit.time=2022-06-16T17\:08\:29+0000
git.dirty=false
git.remote.origin.url=git@github.com\:opensearch-project/security.git
git.tags=
git.total.commit.count=523
```
</p></details>

cliu123 commented:
> Suspicious that a significant portion of the coverage drop comes from whitelist-named files 🤔

The logic and tests for whitelist have not been removed. It's been deprecated, but not removed. Allowlist was introduced, but it included both logic and tests, so it didn't have much impact on [the test coverage](https://github.com/opensearch-project/security/pull/1808#issuecomment-1112763927).

cliu123 commented:
@peternied Great work! Could you move the changes that are not required to standardize the build with OpenSearch plugin? Things like test refactoring, sun_check(non-inclusive terminologies) etc. are not in the scope of this PR. They can be in separate PRs.

peternied commented:
@cliu123 I was able to remove the checkstyle supressions, but the other changes are needed.  Please create comments on individual changes if you think they should be removed.  Calling out the test changes, they are needed because the Parameterized runner cannot be used with RandomizedTest runner, this was preventing these tests from being executed

DarshitChanpura commented:
@peternied Can you please review the failing CI build for java-17?

DarshitChanpura commented on build.gradle:
```diff
@@ -9,79 +9,221 @@
         force 'commons-codec:commons-codec:1.14'
-        force 'org.apache.santuario:xmlsec:2.2.3'
-        force 'org.cryptacular:cryptacular:1.2.4'
-        force 'net.minidev:json-smart:2.4.7'
```
have these libraries been removed to align with core and/or is it because we no longer need them?

DarshitChanpura replied:
okie...got it

DarshitChanpura commented on checkstyle/sun_checks.xml:
```diff
@@ -194,22 +194,14 @@
-  </module>
-
-  <!-- Checks for Headers to not contain old license text -->
-  <module name="RegexpSingleline">
```
why did we remove the old license headers check?

peternied replied:
OpenSearch has its own license header check that I enabled, since we fixed the old license header issues, these specific checks were not needed.

DarshitChanpura replied:
okay..gotcha

DarshitChanpura commented on src/main/java/org/opensearch/security/ssl/util/CertFileProps.java:
```diff
@@ -1,3 +1,14 @@
+ *
+ * Modifications Copyright OpenSearch Contributors. See
+ * GitHub history for details.
+ */
```
should we add license headers to all un-licensed files?

peternied replied:
I've done the minimum work required for the license check from OpenSearch's build system to pass.  I would recommend we create another issue if we want to refine the scope/settings of that tool

DarshitChanpura replied:
okay... if this is required by the core's build system to pass then we do not need to create a separate issue to discuss on this.

DarshitChanpura commented on src/test/java/com/amazon/dlic/auth/http/jwt/HTTPJwtAuthenticatorTest.java:
```diff
@@ -148,15 +145,12 @@ public void testBearerWrongPosition() throws Exception {
     public void testBasicAuthHeader() throws Exception {
         Settings settings = Settings.builder().put("signing_key", BaseEncoding.base64().encode(secretKeyBytes)).build();
         HTTPJwtAuthenticator jwtAuth = new HTTPJwtAuthenticator(settings, null);
-        JwtParser jwtParser = Mockito.spy(JwtParser.class);
```
out of curiosity, why did we remove the parser?

peternied replied:
This test was using an old version of Mockito and the functionality was deprecated, however when I looked at how the test was written this wasn't used so I removed it.

DarshitChanpura replied:
okay

DarshitChanpura commented on src/test/java/org/opensearch/security/TracingTests.java:
```diff
@@ -239,16 +237,6 @@ public void testHTTPTraceNoSource() throws Exception {
     @Test
     public void testHTTPSingle() throws Exception {
-
-        Thread.setDefaultUncaughtExceptionHandler(new UncaughtExceptionHandler() {
```
out of curiosity, why did we remove this handler?

peternied replied:
UncaughtExceptionHandlers can muck with Java internals in ways that make them unpredictable - when we were switched to use Random test runner from OpenSearch's test infrastructure it applied a rule warning about this behavior that is typically bad (but not always).  To get the tests to run without errors I had to remove these calls.

When I reviewed these tests this behavior was not necessary and likely made the tests slower and flaky

DarshitChanpura commented on src/test/java/org/opensearch/security/auditlog/compliance/ComplianceAuditlogTest.java:
```diff
@@ -183,13 +183,12 @@ public void testSourceFilterMsearch() throws Exception {
-        assertNotContains(response, "*exception*");
-        Assert.assertEquals(HttpStatus.SC_OK, response.getStatusCode());
-        Thread.sleep(1500);
+        TestAuditlogImpl.doThenWaitForMessages(() -> {
```
👏

peternied replied:
😄

peternied replied:
However, we still have many places we could apply these changes.  Maybe another time we will do a pass to ensure reliability of tests

DarshitChanpura commented on src/test/java/org/opensearch/security/ccstest/CrossClusterSearchTests.java:
```diff
@@ -80,9 +75,7 @@ public class CrossClusterSearchTests extends AbstractSecurityUnitTest {
-    //default is true
-    @Parameter
-    public boolean ccsMinimizeRoundtrips;
+    protected boolean ccsMinimizeRoundtrips() { return false; };
```
out of curiosity, why are we changing this test to run with `false` only?

DarshitChanpura replied:
okay..so we will still test both conditions right?

DarshitChanpura commented on src/test/java/org/opensearch/security/dlic/rest/api/AccountApiTest.java:
```diff
@@ -29,26 +26,18 @@
 import static org.junit.Assert.assertNotNull;
 import static org.junit.Assert.assertNull;
 import static org.junit.Assert.assertTrue;
-import static org.opensearch.security.OpenSearchSecurityPlugin.LEGACY_OPENDISTRO_PREFIX;
```
out of curiosity, is this removed because we no longer support OpenDistro?

DarshitChanpura commented:
Great contribution @peternied !! This is make development much smoother.

cliu123 commented:
@vrozov had some questions on introducing Randomized Runner and the changes in tests a while ago. Have those questions been answered?
@peternied  Would you check with @vrozov and see if the questions still remain?

peternied commented on src/test/java/com/amazon/dlic/auth/http/saml/MockSamlIdpServer.java:
```diff
@@ -516,7 +515,8 @@ private String createSamlAuthResponse(AuthnRequest authnRequest) {
         KeyEncryptionParameters kek = new KeyEncryptionParameters();
-        kek.setAlgorithm(EncryptionConstants.ALGO_ID_KEYTRANSPORT_RSA15);
+        // Algorithm from https://santuario.apache.org/Java/api/constant-values.html#org.apache.xml.security.utils.EncryptionConstants.ALGO_ID_KEYTRANSPORT_RSA15
+        kek.setAlgorithm("http://www.w3.org/2001/04/xmlenc#rsa-1_5");
```
This is the value that was imported from the library, it allowed that dependency to be removed - faster build teams and cleaner dependency tree for issues like CVEs

DarshitChanpura replied:
oh..okay

peternied commented:
@cliu123 `vrozov` removed himself from reviewing this change, do you have specific concerns that aren't addressed by seeing CI pass?

cliu123 commented:
> @cliu123 `vrozov` removed himself from reviewing this change, do you have specific concerns that aren't addressed by seeing CI pass?

@peternied I know that @vrozov mentioned that Randomized Runner could introduce some regressions for some reason, but @vrozov has more context on this. Would you follow up?

peternied commented:
>I see some test failures in CI. And 1147 tests ran, but only 1111 tests ran before the changes as I commented above. The build system migration change should not add more tests though.

When I compare the number of [tests run](https://github.com/opensearch-project/security/actions/runs/2550007387) against main I see 1146, where as with this change I see 1145.  The number of tests includes retries, so exact numeric matches aren't.  That aside, the associated test coverage is nearly identical.  I don't think there is an issue here - how does this sound to you?

davidlago reviewed (approved):
At this point, I don't think we have enough solid information to block/delay this PR on the suspicion that the Randomized Runner might be causing issues. I support moving forward with this change and addressing any particular issues (if/when confirmed) as they arise. This is an important PR that is long overdue to bring us in line with the rest of the project.

cliu123 commented:
I've dismissed my review.

cliu123 commented on src/main/java/org/opensearch/security/dlic/rest/api/MigrateApiAction.java:
```diff
@@ -27,7 +27,7 @@
 import org.opensearch.action.index.IndexRequest;
 import org.opensearch.action.support.WriteRequest.RefreshPolicy;
-import org.opensearch.action.support.master.AcknowledgedResponse;
+import org.opensearch.action.support.clustermanager.AcknowledgedResponse;
```
This is not required for 2.1.0.

peternied replied:
The build was broken, it seems like this was accidentally introduced in https://github.com/opensearch-project/OpenSearch/pull/3662 into 2.1 - I think its harmless to keep at this point

cliu123 replied:
Including only the required changes in a PR would be better. When regressions happen, it'd be much easier to revert commits with clear PR/commit scope. The broken build was cause by a bug in OpenSearch, and it has been resolved.

cliu123 commented on src/main/java/org/opensearch/security/ssl/util/CertFromFile.java:
```diff
@@ -1,3 +1,14 @@
+ * Modifications Copyright OpenSearch Contributors. See
+ * GitHub history for details.
+ */
+
```
nit: Are License header changes required for this PR? If not, including only the required changes in a PR would be better. When regressions happen, it'd be much easier to revert commits with clear PR/commit scope.

peternied replied:
License header checks are required by default in the OpenSearch build scripts, seemed like a good opportunity to just do it instead of attempting a work around.

cliu123 commented on src/test/java/com/amazon/dlic/auth/http/jwt/HTTPJwtAuthenticatorTest.java:
```diff
@@ -148,15 +145,12 @@ public void testBearerWrongPosition() throws Exception {
         Settings settings = Settings.builder().put("signing_key", BaseEncoding.base64().encode(secretKeyBytes)).build();
         HTTPJwtAuthenticator jwtAuth = new HTTPJwtAuthenticator(settings, null);
-        JwtParser jwtParser = Mockito.spy(JwtParser.class);
-        FieldSetter.setField(jwtAuth, HTTPJwtAuthenticator.class.getDeclaredField("jwtParser"), jwtParser);
```
Same comment on PR scope, looks like these test changes are not required for this PR either.

peternied replied:
Using OpenSearch's test system it imported a newer version of Mockito that no longer supports `FieldSetter`, tests wouldn't compile without this change.

cliu123 commented on src/test/java/org/opensearch/security/ConfigTests.java:
```diff
@@ -76,26 +77,27 @@ public void testMigrate() throws Exception {
 
         check("./legacy/securityconfig_v6/action_groups.yml", CType.ACTIONGROUPS);
-        check("./securityconfig/action_groups.yml", CType.ACTIONGROUPS);
+        check("./action_groups.yml", CType.ACTIONGROUPS);
```
Are these changes required for this PR?

peternied replied:
See https://github.com/opensearch-project/security/pull/1888#discussion_r906241050

cliu123 commented on src/test/java/org/opensearch/security/SecurityAdminTests.java:
```diff
@@ -89,8 +88,7 @@ public void testSecurityAdminInvalidCert() throws Exception {
         argsAsList.add(clusterInfo.clustername);
-        argsAsList.add("-cd");
-        argsAsList.add(new File("src/test/resources/").getAbsolutePath());
+        addDirectoryPath(argsAsList, TEST_RESOURCE_ABSOLUTE_PATH);
```
Same comment on PR scope, looks like these test changes are not required for this PR either.

peternied replied:
We had many tests that manipulate the file system, under the OpenSearch test runner these paths have changed.  I mitigated as much as possible to prevent code change but some tests needed individual attention

cliu123 commented on src/test/java/org/opensearch/security/dlic/dlsfls/DlsFlsCrossClusterMinimalRoundtripSearchTests.java:
```diff
@@ -0,0 +1,17 @@
+/*
```
I see a few new test files have been added in the PR. Why are they needed?

peternied replied:
I had to remove use of the Parameterized test runner since Randomized runner was in use, these new files extend the existing tests to keep the test coverage intact with minimal additional code.
